### PR TITLE
Horizontal split view for Filesystem Dock

### PR DESCRIPTION
**Pull request by gamedevishard**
filesystem_dock

filesystem_dock

### CODE REVIEW

File: editor/filesystem_dock.cpp [hunk 1]
```diff
@@ -3382,17 +3406,23 @@ FileSystemDock::FileSystemDock() {
...
 
-	split_box = memnew(VSplitContainer);
-	split_box->set_v_size_flags(SIZE_EXPAND_FILL);
-	add_child(split_box);
+	vsplit_box = memnew(VSplitContainer);
+	vsplit_box->set_v_size_flags(SIZE_EXPAND_FILL);
+	add_child(vsplit_box);
+	split_box = vsplit_box;
+
+	hsplit_box = memnew(HSplitContainer);
```
Comment (KoBeWi) [inline editor/filesystem_dock.cpp:3414]: You could just use `SplitContainer` and toggle `vertical` I think.

Comment (gamedevishard) [reply editor/filesystem_dock.cpp]: @KoBeWi Thanks for the review
Fixed

File: editor/filesystem_dock.cpp [hunk 2]
```diff
@@ -426,22 +426,34 @@ void FileSystemDock::set_display_mode(DisplayMode p_display_mode) {
...
+				toolbar2_hbc->show();
 
 				_update_tree(get_uncollapsed_paths());
 				file_list_vb->hide();
 				break;
 
-			case DISPLAY_MODE_SPLIT:
+			case DISPLAY_MODE_HSPLIT:
+			case DISPLAY_MODE_VSPLIT:
+				bool isVertical = display_mode == DISPLAY_MODE_VSPLIT;
```
Comment (AThousandShips) [inline editor/filesystem_dock.cpp:442]:
```suggestion
				bool is_vertical = display_mode == DISPLAY_MODE_VSPLIT;
```
Use `snake_case` in c++

Comment (nakoff) [reply editor/filesystem_dock.cpp]: Done

Comment (gamedevishard) [reply editor/filesystem_dock.cpp]: Thanks, Done.

File: editor/filesystem_dock.cpp [hunk 3]
```diff
@@ -471,7 +483,7 @@ void FileSystemDock::_notification(int p_what) {
 			EditorResourcePreview::get_singleton()->connect("preview_invalidated", callable_mp(this, &FileSystemDock::_preview_invalidated));
 
 			button_reload->set_icon(get_theme_icon(SNAME("Reload"), SNAME("EditorIcons")));
-			button_toggle_display_mode->set_icon(get_theme_icon(SNAME("Panels2"), SNAME("EditorIcons")));
+			button_toggle_display_mode->set_icon(get_theme_icon(SNAME("Panels1"), SNAME("EditorIcons")));
```
Comment (KoBeWi) [inline editor/filesystem_dock.cpp:486]: This icon is overridden in `_update_display_mode()`, so the line can be removed.

File: editor/filesystem_dock.cpp [hunk 4]
```diff
@@ -558,7 +570,7 @@ void FileSystemDock::_notification(int p_what) {
 		case EditorSettings::NOTIFICATION_EDITOR_SETTINGS_CHANGED: {
 			// Update icons.
 			button_reload->set_icon(get_theme_icon(SNAME("Reload"), SNAME("EditorIcons")));
-			button_toggle_display_mode->set_icon(get_theme_icon(SNAME("Panels2"), SNAME("EditorIcons")));
+			button_toggle_display_mode->set_icon(get_theme_icon(SNAME("Panels1"), SNAME("EditorIcons")));
```
Comment (KoBeWi) [inline editor/filesystem_dock.cpp:573]: This will assign wrong icon when settings change.

File: editor/filesystem_dock.cpp [hunk 5]
```diff
@@ -426,22 +426,34 @@ void FileSystemDock::set_display_mode(DisplayMode p_display_mode) {
...
+			case DISPLAY_MODE_HSPLIT:
+			case DISPLAY_MODE_VSPLIT:
+				bool is_vertical = display_mode == DISPLAY_MODE_VSPLIT;
+				StringName icon = is_vertical ? SNAME("Panels2") : SNAME("Panels2Alt");
+				split_box->set_vertical(is_vertical);
+				button_toggle_display_mode->set_icon(get_theme_icon(icon, SNAME("EditorIcons")));
+
+				if (tree->get_parent() != split_box) {
+					tree->get_parent()->remove_child(tree);
+					split_box->add_child(tree);
```
Comment (KoBeWi) [inline editor/filesystem_dock.cpp:449]:
```suggestion
					tree->reparent(split_box, false);
```

File: editor/filesystem_dock.cpp [hunk 6]
```diff
@@ -426,22 +426,34 @@ void FileSystemDock::set_display_mode(DisplayMode p_display_mode) {
...
+				}
+
+				if (file_list_vb->get_parent() != split_box) {
+					file_list_vb->get_parent()->remove_child(file_list_vb);
+					split_box->add_child(file_list_vb);
```
Comment (KoBeWi) [inline editor/filesystem_dock.cpp:454]: Same as above.

File: editor/filesystem_dock.cpp (hunk 5 above)
Comment (KoBeWi) [reply editor/filesystem_dock.cpp:449]: Although why is this even necessary? The tree is always under `split_box`.

File: editor/filesystem_dock.cpp [hunk 7]
```diff
@@ -426,22 +426,34 @@ void FileSystemDock::set_display_mode(DisplayMode p_display_mode) {
...
 				_update_tree(get_uncollapsed_paths());
 				file_list_vb->hide();
 				break;
 
-			case DISPLAY_MODE_SPLIT:
+			case DISPLAY_MODE_HSPLIT:
+			case DISPLAY_MODE_VSPLIT:
+				bool is_vertical = display_mode == DISPLAY_MODE_VSPLIT;
+				StringName icon = is_vertical ? SNAME("Panels2") : SNAME("Panels2Alt");
+				split_box->set_vertical(is_vertical);
```
Comment (KoBeWi) [inline editor/filesystem_dock.cpp:444]:
```suggestion
				split_box->set_vertical(is_vertical);
				StringName icon = is_vertical ? SNAME("Panels2") : SNAME("Panels2Alt");
```

File: editor/filesystem_dock.cpp [hunk 8]
```diff
@@ -426,22 +426,34 @@ void FileSystemDock::set_display_mode(DisplayMode p_display_mode) {
 void FileSystemDock::_update_display_mode(bool p_force) {
 	// Compute the new display mode.
 	if (p_force || old_display_mode != display_mode) {
-		button_toggle_display_mode->set_pressed(display_mode == DISPLAY_MODE_SPLIT);
 		switch (display_mode) {
 			case DISPLAY_MODE_TREE_ONLY:
+				button_toggle_display_mode->set_icon(get_theme_icon(SNAME("Panels1"), SNAME("EditorIcons")));
```
Comment (KoBeWi) [inline editor/filesystem_dock.cpp:431]:
```suggestion
				button_toggle_display_mode->set_icon(get_editor_theme_icon(SNAME("Panels1")));
```
This was just merged, so you should use the new method.

**Review by KoBeWi — APPROVED** [review]
Seems fine now. One concern is that when you switch between horizontal and vertical mode, the split will be wrong. Maybe we should keep it separately for 2 modes

**Comment by gamedevishard** [comment]
> Seems fine now. One concern is that when you switch between horizontal and vertical mode, the split will be wrong. Maybe we should keep it separately for 2 modes

Check now please

File: editor/filesystem_dock.cpp (hunk 7 above)
Comment (gamedevishard) [reply editor/filesystem_dock.cpp]: fixed

**Comment by KoBeWi** [comment]
The split seems to be not remembered between sessions.
Set vertical split size and horizontal split size. Restart while the dock is in horizontal mode. The horizontal split will reset, while the vertical split will use the horizontal value. I didn't check the implementation, but they need to be stored in separate values.

**Comment by akien-mga** [comment]
Thanks! And congrats for your first merged Godot contribution :tada:

**Pull request merged**